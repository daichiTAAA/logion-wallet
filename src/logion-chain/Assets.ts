import { ApiPromise } from '@polkadot/api';
import { StorageKey, Option } from '@polkadot/types';
import {
    AssetId as PolkadotAssetId,
    AssetDetails as PolkadotAsset,
    AssetMetadata as PolkadotAssetMetadata,
    Call
} from '@polkadot/types/interfaces';
import { randomBytes } from 'crypto';
import BN from 'bn.js';
import {
    AssetId as AssetIdType,
    AssetBalance as AssetBalanceType,
    AssetMetadata as AssetMetadataType,
    Asset as AssetType,
    AssetWithBalance as AssetWithBalanceType,
} from './Types';

import {
    SignAndSendCallback,
    ErrorCallback,
    signAndSend,
    Unsubscriber
} from './Signature';

export const DEFAULT_ASSETS_DECIMALS = 18;

export interface AssetCreationParameters {
    signerId: string,
    callback: SignAndSendCallback,
    errorCallback: ErrorCallback,
    api: ApiPromise,
}

export type AssetId = AssetIdType;

export interface AssetCreationResult {
    assetId: AssetId,
    unsubscriber: Unsubscriber,
}

export async function createAsset(parameters: AssetCreationParameters): Promise<AssetCreationResult> {
    const {
        api,
        signerId,
        callback,
        errorCallback,
    } = parameters;

    const assetId = await nextAssetId(api);
    const assetCreate: any = api.tx.assets.create; // circumvent typing in order to add missing argument
    const unsubscriber = signAndSend({
        signerId,
        submittable: assetCreate(assetId, signerId, 10, 1),
        callback,
        errorCallback,
    });

    return {
        assetId,
        unsubscriber
    };
}

async function nextAssetId(api: ApiPromise): Promise<AssetId> {
    const BYTES_IN_ASSET_ID = 64 / 8;
    const value = randomBytes(BYTES_IN_ASSET_ID);
    let candidate = new BN(value.toString('hex'), 16);
    let candidateUsed = await assetExists(api, candidate);

    let iteration = 0;
    const MAX_ITERATIONS = 10;
    while(candidateUsed && iteration < MAX_ITERATIONS) {
        candidate = new BN(value.toString('hex'), 16);
        candidateUsed = await assetExists(api, candidate);
        ++iteration;
    }

    if(candidateUsed) {
        throw new Error("Unable to select an unused ID");
    } else {
        return candidate;
    }
}

async function assetExists(api: ApiPromise, id: AssetId): Promise<boolean> {
    const asset = await api.query.assets.asset(id);
    return asset.isSome;
}

export type AssetMetadata = AssetMetadataType;

export interface SetAssetMetadataParameters {
    signerId: string,
    callback: SignAndSendCallback,
    errorCallback: ErrorCallback,
    api: ApiPromise,
    assetId: AssetId,
    metadata: AssetMetadata,
}

export function setAssetMetadata(parameters: SetAssetMetadataParameters): Unsubscriber {
    const {
        api,
        signerId,
        callback,
        errorCallback,
        assetId,
        metadata,
    } = parameters;
    return signAndSend({
        signerId,
        submittable: api.tx.assets.setMetadata(assetId, metadata.name, metadata.symbol, metadata.decimals),
        callback,
        errorCallback,
    });
}
export type AssetBalance = AssetBalanceType;

export function balanceFromAmount(tokens: number, decimals: number): AssetBalance {
    let tokensBN = new BN(tokens);
    let decimalsBN = new BN(decimals);
    let ten = new BN(10);
    return tokensBN.mul(ten.pow(decimalsBN));
}

export interface MintParameters {
    signerId: string,
    beneficiary: string,
    callback: SignAndSendCallback,
    errorCallback: ErrorCallback,
    api: ApiPromise,
    assetId: AssetId,
    amount: AssetBalance
}

export function mintTokens(parameters: MintParameters): Unsubscriber {
    const {
        api,
        signerId,
        beneficiary,
        callback,
        errorCallback,
        assetId,
        amount,
    } = parameters;
    return signAndSend({
        signerId,
        submittable: api.tx.assets.mint(assetId, beneficiary, amount),
        callback,
        errorCallback,
    });
}

export function tokensFromBalance(balance: AssetBalance, decimals: number): string {
    let balanceBN = new BN(balance);
    let exponent = new BN(decimals);
    let ten = new BN(10);
    return balanceBN.div(ten.pow(exponent)).toString();
}

export interface AssetBalanceParameters {
    api: ApiPromise,
    assetId: AssetId,
    decimals: number,
    address: string,
}

export async function assetBalance(parameters: AssetBalanceParameters): Promise<string> {
    const {
        api,
        assetId,
        decimals,
        address,
    } = parameters;
    const balance = await api.query.assets.account(assetId, address);
    return tokensFromBalance(balance.balance, decimals);
}

export interface GetAssetsParameters {
    api: ApiPromise,
}

export type Asset = AssetType;

export async function getAssets(parameters: GetAssetsParameters): Promise<Asset[]> {
    const {
        api,
    } = parameters;
    const assetIds = extractAssetIds(await api.query.assets.asset.keys());
    const assetDetails: Option<PolkadotAsset>[] = await Promise.all(assetIds.map(assetId => api.query.assets.asset(assetId)));
    const metadata: PolkadotAssetMetadata[] = await Promise.all(assetIds.map(assetId => api.query.assets.metadata(assetId)));
    return assetIds.map((assetId, index) => {
        return {
            assetId,
            issuer: assetDetails[0].unwrap().issuer.toString(),
            metadata: {
                name: metadata[index].name.toUtf8(),
                symbol: metadata[index].symbol.toUtf8(),
                decimals: metadata[index].decimals.toNumber(),
            }
        };
    });
}

function extractAssetIds(keys: StorageKey<[PolkadotAssetId]>[]): AssetId[] {
    return keys.map(({ args: [assetId] }) => assetId);
}

export interface AccountBalanceParameters {
    api: ApiPromise,
    accountId: string,
    assets?: Asset[],
}

export type AssetWithBalance = AssetWithBalanceType;

export async function accountBalance(parameters: AccountBalanceParameters): Promise<AssetWithBalance[]> {
    const {
        api,
        accountId,
        assets,
    } = parameters;
    

    let theAssets;
    if(assets !== undefined) {
        theAssets = assets;
    } else {
        theAssets = await getAssets({api: api!});
    }
    const balances: string[] = await Promise.all(theAssets.map(asset => assetBalance({
        api,
        assetId: asset.assetId,
        decimals: asset.metadata.decimals,
        address: accountId
    })));
    return theAssets.map((asset, index) => ({
        asset,
        balance: balances[index],
    }));
}

export interface BuildTransferCallParameters {
    api: ApiPromise,
    assetId: AssetId,
    target: string,
    amount: AssetBalance,
}

export function buildTransferCall(parameters: BuildTransferCallParameters): Call {
    const {
        api,
        assetId,
        target,
        amount,
    } = parameters;

    return api.createType('Call', api.tx.assets.transfer(assetId, target, amount));
}
