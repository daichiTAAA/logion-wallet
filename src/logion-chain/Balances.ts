import { ApiPromise } from '@polkadot/api';
import { PrefixedNumber, ScientificNumber, convertToPrefixed, NONE, ATTO } from "./numbers";
import { ExtrinsicSubmissionParameters, Unsubscriber, signAndSend } from './Signature';
import { Call } from "@polkadot/types/interfaces";
import { SubmittableExtrinsic } from "@polkadot/api/submittable/types";

const LGNT_DECIMALS = 18;
export const LGNT_SMALLEST_UNIT = ATTO;
export const SYMBOL = "LGNT";

export interface GetAccountDataParameters {
    api: ApiPromise,
    accountId: string,
}

export interface AccountData {
    available: string,
    reserved: string,
    total: string,
}

export async function getAccountData(parameters: GetAccountDataParameters): Promise<AccountData> {
    const {
        api,
        accountId,
    } = parameters;

    const accountData = await api.query.system.account(accountId);

    return {
        available: accountData.data.free.toString(),
        reserved: accountData.data.reserved.toString(),
        total: accountData.data.free.add(accountData.data.reserved).toString(),
    };
}
export interface Coin {
    id: string,
    name: string,
    iconId: string,
    iconType: 'svg' | 'png',
    symbol: string,
}

export interface CoinBalance {
    coin: Coin,
    balance: PrefixedNumber,
    level: number,
}

export async function getBalances(parameters: GetAccountDataParameters): Promise<CoinBalance[]> {
    const {
        api,
        accountId,
    } = parameters;

    const data = await getAccountData({
        api: api!,
        accountId,
    });

    const logAvailable = scientificLogBalance(data.total).optimizeScale(3);
    const logPrefixedAvailable = convertToPrefixed(logAvailable);
    const logLevel = logAvailable.divideBy(ARTIFICIAL_MAX_BALANCE).toNumber();

    return [
        buildCoinBalance('lgnt', logPrefixedAvailable, logLevel),
        buildCoinBalance('dot', new PrefixedNumber("0", NONE), 1)
    ];
}

export function scientificLogBalance(tokens: string): ScientificNumber {
    return new ScientificNumber(tokens, -LGNT_DECIMALS).optimizeScale(3);
}

export function prefixedLogBalance(tokens: string): PrefixedNumber {
    const scientific = scientificLogBalance(tokens);
    return convertToPrefixed(scientific);
}

const ARTIFICIAL_MAX_BALANCE = scientificLogBalance("100");

function buildCoinBalance(coinId: string, balance: PrefixedNumber, level: number): CoinBalance {
    const coin = getCoin(coinId);
    return {
        coin,
        balance,
        level,
    }
}

export function getCoin(coinId: string): Coin {
    if(coinId === 'dot') {
        return {
            id: 'dot',
            name: 'Polkadot',
            iconId: 'dot',
            iconType: 'png',
            symbol: 'DOT',
        };
    } else if(coinId === "lgnt") {
        return {
            id: 'lgnt',
            name: 'logion Token',
            iconId: 'lgnt',
            iconType: 'svg',
            symbol: SYMBOL,
        };
    } else {
        throw new Error(`Unsupported coin ${coinId}`);
    }
}

export interface TransferParameters extends ExtrinsicSubmissionParameters, BuildTransferCallParameters {
}

export function transfer(parameters: TransferParameters): Unsubscriber {
    const {
        signerId,
        callback,
        errorCallback,
    } = parameters;

    return signAndSend({
        signerId,
        submittable: transferSubmittable(parameters),
        callback,
        errorCallback,
    });
}

export interface BuildTransferCallParameters {
    api: ApiPromise;
    destination: string;
    amount: PrefixedNumber;
}

function transferSubmittable(parameters: BuildTransferCallParameters): SubmittableExtrinsic<'promise'> {
    const {
        api,
        destination,
        amount
    } = parameters;
    return api.tx.balances.transfer(destination, amount.convertTo(LGNT_SMALLEST_UNIT).coefficient.unnormalize())
}

export function buildTransferCall(parameters: BuildTransferCallParameters): Call {
    return parameters.api.createType('Call', transferSubmittable(parameters))
}

export async function estimateFee(parameters: BuildTransferCallParameters): Promise<PrefixedNumber> {
    const submittable = transferSubmittable(parameters);
    const paymentInfo = await submittable.paymentInfo(parameters.destination);
    return new PrefixedNumber(paymentInfo.partialFee.toString(), LGNT_SMALLEST_UNIT);
}
