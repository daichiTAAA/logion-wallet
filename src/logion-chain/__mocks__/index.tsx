import { Extension } from '../Keys';

const LogionChainContextProvider = (props: any) => null;

let context = {
    apiState: 'CONNECT_INIT',
    injectedAccounts: []
};

export function setContextMock(value: any) {
    context = value;
}

export function useLogionChain() {
    return context;
}

let extensionAvailable = false;

export function setExtensionAvailable(value: boolean) {
    extensionAvailable = value;
}

export function isExtensionAvailable() {
    return extensionAvailable;
}

let recommendedExtensionValue: Extension | null = null;

export function setRecommendedExtension(value: Extension | null) {
    recommendedExtensionValue = value;
}

export function recommendedExtension(): Extension | null {
    return recommendedExtensionValue;
}

export {
    LogionChainContextProvider,
};
