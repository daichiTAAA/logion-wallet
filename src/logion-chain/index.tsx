// Imports
import { KeyringPair as PolkadotKeyringPair } from '@polkadot/keyring/types';
import {
    Block as PolkadotBlock,
    Hash as PolkadotHash,
    Header as PolkadotHeader,
    SignedBlock as PolkadotSignedBlock,
    EventRecord as PolkadotEventRecord,
    Extrinsic as PolkadotExtrinsic,
} from '@polkadot/types/interfaces';
import keyring from '@polkadot/ui-keyring';

import {
    LogionChainContextProvider,
    useLogionChain,

    ExtrinsicFetchSpecification as ContextExtrinsicFetchSpecification,
    ApiState as ContextApiState,
    NodeMetadata as ContextNodeMetadata,
} from './LogionChainContext';

import {
    signAndSend,
    replaceUnsubscriber,

    SignatureParameters as SignatureSignatureParameters,
    Unsubscriber as SignatureUnsubscriber,
    SignAndSendCallback as SignatureSignAndSendCallback,
} from './Signature';

import {
    getTimestamp,
    fetchExtrinsics,
    hashEquals,

    ExtrinsicsAndHead as BlocksExtrinsicsAndHead,
} from './Blocks';

import {
    isExtensionAvailable,
    recommendedExtension
} from './Keys';

// Re-exports
export {
    useLogionChain,
    LogionChainContextProvider,
    getTimestamp,
    fetchExtrinsics,
    hashEquals,
    signAndSend,
    replaceUnsubscriber,
    isExtensionAvailable,
    recommendedExtension,
    keyring,
};

export type Unsubscriber = SignatureUnsubscriber;
export type SignatureParameters = SignatureSignatureParameters;
export type ExtrinsicFetchSpecification = ContextExtrinsicFetchSpecification;

export type ExtrinsicsAndHead = BlocksExtrinsicsAndHead;

export type KeyringPair = PolkadotKeyringPair;
export type Hash = PolkadotHash;
export type SignedBlock = PolkadotSignedBlock;
export type Header = PolkadotHeader;
export type Block = PolkadotBlock;
export type EventRecord = PolkadotEventRecord;
export type Extrinsic = PolkadotExtrinsic;
export type SignAndSendCallback = SignatureSignAndSendCallback;
export type ApiState = ContextApiState;
export type NodeMetadata = ContextNodeMetadata;
