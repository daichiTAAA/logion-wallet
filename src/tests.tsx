import { ReactElement } from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import reactRenderer, { ReactTestRenderer, ReactTestRendererJSON, act as reactAct } from 'react-test-renderer';

export function render(element: ReactElement): null | ReactTestRendererJSON | ReactTestRendererJSON[] {
    const component = reactRenderer.create(element);
    return component.toJSON();
}

export function renderer(element: ReactElement): ReactTestRenderer {
    return reactRenderer.create(element);
}

export function shallowRender<E extends ReactElement>(element: ReactElement): E {
    const renderer = createRenderer();
    renderer.render(element);
    return renderer.getRenderOutput();
}

export function mockAccount(address: string, name: string) {
    return {
        address,
        meta: {
            name
        }
    };
}

export { reactAct as act };
