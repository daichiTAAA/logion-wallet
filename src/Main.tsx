import React from 'react';
import { useLogionChain, isExtensionAvailable } from './logion-chain';

import Wallet from './Wallet';
import Loader from './Loader';
import InstallExtension from './InstallExtension';
import CreateAccount from './CreateAccount';

export default function Main() {
    const { injectedAccounts } = useLogionChain();

    if(isExtensionAvailable()) {
        if(injectedAccounts === null) {
            return <Loader text="Loading accounts from extension..." />;
        } else {
            if(injectedAccounts.length > 0) {
                return <Wallet />;
            } else {
                return <CreateAccount />;
            }
        }
    } else {
        return <InstallExtension />;
    }
}
