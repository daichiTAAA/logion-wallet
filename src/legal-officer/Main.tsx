import React from 'react';

import { useLogionChain } from '../logion-chain';

import { LegalOfficerContextProvider } from './LegalOfficerContext';
import ContextualizedWallet from './ContextualizedWallet';

export default function LegalOfficerWallet() {
    const { injectedAccounts } = useLogionChain();

    if(injectedAccounts === null) {
        return null;
    }

    return (
        <LegalOfficerContextProvider>
            <ContextualizedWallet />
        </LegalOfficerContextProvider>
    );
}
