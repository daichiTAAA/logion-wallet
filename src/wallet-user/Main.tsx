import React from 'react';

import { useLogionChain } from '../logion-chain';

import { UserContextProvider } from "./UserContext";
import ContextualizedWallet from './ContextualizedWallet';

export default function Wallet() {
    const { injectedAccounts } = useLogionChain();

    if(injectedAccounts === null) {
        return null;
    }

    return (
        <UserContextProvider>
            <ContextualizedWallet />
        </UserContextProvider>
    );
}
