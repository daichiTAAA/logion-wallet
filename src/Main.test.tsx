jest.mock('./logion-chain');
jest.mock('./config');

import React from 'react';
import Main from './Main';
import { setContextMock, setExtensionAvailable } from './logion-chain';
import { shallowRender, mockAccount } from './tests';

test('Given no extension, when rendering, then show install', () => {
    setExtensionAvailable(false);

    const tree = shallowRender(<Main />);

    expect(tree).toMatchSnapshot();
});

test('Given extension and loading accounts, when rendering, then show loader', () => {
    setExtensionAvailable(true);

    const tree = shallowRender(<Main />);

    expect(tree).toMatchSnapshot();
});

test('Given extension and no account, when rendering, then show create account', () => {
    setContextMock({
        injectedAccounts: []
    });
    setExtensionAvailable(true);

    const tree = shallowRender(<Main />);

    expect(tree).toMatchSnapshot();
});

test('Given extension and one account, when rendering, then show wallet', () => {
    setContextMock({
        apiState: 'READY',
        injectedAccounts: [
            mockAccount("address", "Account name")
        ]
    });
    setExtensionAvailable(true);

    const tree = shallowRender(<Main />);

    expect(tree).toMatchSnapshot();
});
