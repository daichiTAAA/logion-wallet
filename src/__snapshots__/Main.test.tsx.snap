// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given extension and loading accounts, when rendering, then show loader 1`] = `
<Loader
  text="Loading accounts from extension..."
/>
`;

exports[`Given extension and no account, when rendering, then show create account 1`] = `<CreateAccount />`;

exports[`Given extension and one account, when rendering, then show wallet 1`] = `<Wallet />`;

exports[`Given no extension, when rendering, then show install 1`] = `<InstallExtension />`;
